[MRG+1] fix #6101 GradientBoosting decision_function for sparse inputs

olologin opened the pull request:
Fix for issue #6101 
Please make suggestions.

olologin committed:
small refactoring

olologin committed:
Now GradientBoosting predict method works with csr_matrix

olologin committed:
Description updated

olologin committed:
Tests for GradientBoosting updated

aflaxman commented:
This makes my example #6101 work. Thanks!

jmschrei commented:
This looks mostly good to me. You should squash the commits as well. @glouppe can you take a look?

amueller commented:
Have you check the prediction speed for single samples? There is a benchmark in the benchmarks folder, I think.

olologin commented:
@amueller Hmm, it shouldn't slow down anything, because this PR only adds prediction functionality for sparse matricies. Also, can you point me at that benchmark? I can't find anything related to GradientBoosting in benchmark folder.

olologin commented:
Sorry for so late response, could someone check it? @amueller, @glouppe 

    20 newsgroups
    =============
    X_train.shape = (11314, 130107)
    X_train density = 0.001214353154362896
    y_train (11314,)
    X_test (3500, 130107)
    X_test.format = csr
    X_test.dtype = float32
    y_test (3500,)

    Classifier Training
    ===================
    Training GradientBoostingClassifier_100_trees ...
    
    done

    Classification performance:
    ===========================

    Classifier   test-time Accuracy
    --------------------------------------------
    Dense GradientBoostingClassifier_100_trees  0.9191s     0.7454
    Sparse GradientBoostingClassifier_100_trees  1.0969s     0.7454 


I've made some benchmark based on bench_20_newsgroups.py:

    from __future__ import print_function, division
    from time import time
    import numpy as np
    
    from sklearn.datasets import fetch_20newsgroups_vectorized
    from sklearn.metrics import accuracy_score
    from sklearn.utils.validation import check_array
    
    from sklearn.ensemble import GradientBoostingClassifier
    
    ESTIMATORS = {
        "GradientBoostingClassifier_100_trees": GradientBoostingClassifier(n_estimators=100)#,
        #"GradientBoostingClassifier_200_trees": GradientBoostingClassifier(n_estimators=200)
    }
    
    data_train = fetch_20newsgroups_vectorized(subset="train")
    data_test = fetch_20newsgroups_vectorized(subset="test")
    X_train_sp = check_array(data_train.data, dtype=np.float32,
                          accept_sparse="csc")
    checked_test = check_array(data_test.data, dtype=np.float32, accept_sparse="csr")
    X_test_sp = checked_test[:3500, :]
    y_train_sp = data_train.target
    y_test_sp = data_test.target[:3500]
    
    X_test_dense = X_test_sp.todense()
    
    print("20 newsgroups")
    print("=============")
    print("X_train.shape = {0}".format(X_train_sp.shape))
    print("X_train density = {0}"
          "".format(X_train_sp.nnz / np.product(X_train_sp.shape)))
    print("y_train {0}".format(y_train_sp.shape))
    print("X_test {0}".format(X_test_sp.shape))
    print("X_test.format = {0}".format(X_test_sp.format))
    print("X_test.dtype = {0}".format(X_test_sp.dtype))
    print("y_test {0}".format(y_test_sp.shape))
    print()
    
    print("Classifier Training")
    print("===================")
    accuracy, test_time = {}, {}
    for name, clf in ESTIMATORS.items():
        try:
            clf.set_params(random_state=0)
        except (TypeError, ValueError):
            pass
    
        print("Training %s ... " % name, end="")
        clf.fit(X_train_sp, y_train_sp)
        t0 = time()
        y_pred = clf.predict(X_test_sp)
        test_time["Sparse " + name] = time() - t0
        accuracy["Sparse " + name] = accuracy_score(y_test_sp, y_pred)
    
        t0 = time()
        y_pred = clf.predict(X_test_dense)
        test_time["Dense " + name] = time() - t0
        accuracy["Dense " + name] = accuracy_score(y_test_sp, y_pred)
        
        print("done")
    
    print()
    print("Classification performance:")
    print("===========================")
    print()
    print("%s %s %s" % ("Classifier  ", "test-time", "Accuracy"))
    print("-" * 44)
    for name in sorted(accuracy, key=accuracy.get):
        print("%s %s %s" % (name.ljust(16),
                            ("%.4fs" % test_time[name]).center(10),
                            ("%.4f" % accuracy[name]).center(10)))
    



Classifier Training
===================
Training GradientBoostingClassifier_100_trees ... `

l3link commented:
This fix is incredibly useful for very sparse data sets ( >95% 0 values).  Converting a medium sized data set with a 60k x 3k matrix from dense to sparse reduces training time from hours to minutes (on a c3.8xlarge AWS single machine).  Any chance we can get this merged into the next release?  Or at least into develop? I'm getting annoyed building this from source for the last 2 months.

glouppe commented:
From a quick read, this looks good, besides some minor cosmit issues.

glouppe commented on sklearn/ensemble/_gradient_boosting.pyx:
```diff
@@ -27,24 +36,19 @@ from numpy import bool as np_bool
+                                                     Py_ssize_t K,
+                                                     Py_ssize_t n_samples,
+                                                     Py_ssize_t n_features,
+                                                     float64 *out):
```
Should be indented with `DTYPE_t`, not with the opening parenthesis.

glouppe commented on sklearn/ensemble/_gradient_boosting.pyx:
```diff
@@ -84,22 +88,100 @@ cdef void _predict_regression_tree_inplace_fast(DTYPE_t *X,
                 node = root_node + node.right_child
         out[i * K + k] += scale * value[node - root_node]
 
+@cython.nonecheck(False)
```
By default, `nonecheck=False`. This is decorator can therefore be removed.

jaquesgrobler commented:
I had a quick read-through.. Apart from @glouppe 's comments above, this looks great.
Very useful fix this. I'm +1 for merging this once last points are addressed :+1:

jmschrei commented:
This seems extremely useful. Should be merged as soon as comments are addressed.

jnothman commented:
@olologin or whoever does the merge should remember to add a what's new entry.

olologin committed:
What's new entry added

olologin committed:
cython.nonecheck(False) removed, Indentation in _predict_regression_tree_inplace_fast_dense fixed

olologin commented:
@glouppe , @jaquesgrobler , @jnothman .

Fixed, AppVeyour build fails, but seems it's not my fault.

olologin commented:
ping @jnothman

amueller commented:
please rebase.

amueller commented on doc/whats_new.rst:
```diff
@@ -215,6 +215,10 @@ Enhancements
      By `Sebastian Säger`_ and `YenChen Lin`_.
 
+   - :class:`ensemble.GradientBoostingClassifier` and :class:`ensemble.GradientBoostingRegressor`
+     now support sparse input for ``predict`` method.
```
*the predict method

amueller commented:
@glouppe @jnothman does this have your +1 and can be merged, or should I review?

jnothman commented:
I've not looked at this yet.

jnothman commented:
Btw, at a skim this looks good, but I'd like to look through it more closely.

jnothman commented on doc/whats_new.rst:
```diff
@@ -230,6 +230,10 @@ Enhancements
      By `Sebastian Säger`_ and `YenChen Lin`_.
 
+   - :class:`ensemble.GradientBoostingClassifier` and :class:`ensemble.GradientBoostingRegressor`
+     now support sparse input for the ``predict`` method.
```
might as well say for "prediction" to include other prediction methods.

jnothman commented on sklearn/ensemble/_gradient_boosting.pyx:
```diff
@@ -84,22 +88,99 @@ cdef void _predict_regression_tree_inplace_fast(DTYPE_t *X,
+
+    The function assumes that the ndarray that wraps ``X`` is csr_matrix.
+    """
+    cdef np.ndarray[ndim=1, dtype=DTYPE_t] X_data_ndarray = X.data
```
I think all these ephemeral variables make the code harder to read. We can either do the casting in one line, or accept these as parameters to the function.

Yes, I know you got it from `tree.pyx`.

jnothman commented on sklearn/ensemble/_gradient_boosting.pyx:
```diff
@@ -84,22 +88,99 @@ cdef void _predict_regression_tree_inplace_fast(DTYPE_t *X,
+        for stage_idx in range(n_stages):
+            # Cycle through all trees
+            for output_idx in range(n_outputs):
+                root_node = (<Tree> estimators[stage_idx, output_idx].tree_).nodes
```
I note that this is native Python work, but probably negligible. If we find  slowness, we might consider materialising an array of `Tree`s outside the loop over samples.

Also: pep8 line length

olologin replied:
@jnothman You mean casting pointer to Tree and dereferencing it?
And you want me to change this to some 2d array (n_stages, n_outputs) with casted to Tree pointers, which I'll fill in advance?

olologin replied:
@jnothman, or you mean cycling through all tress, and inside of it cycling through all of samples? In this case I doubt it can be faster.

jnothman commented on sklearn/ensemble/_gradient_boosting.pyx:
```diff
@@ -84,22 +88,99 @@ cdef void _predict_regression_tree_inplace_fast(DTYPE_t *X,
+                        node = root_node + node.left_child
+                    else:
+                        node = root_node + node.right_child
+                out_ptr[sample_idx * n_outputs + output_idx] += scale * value[node - root_node]
```
pep8 line length

jnothman commented on sklearn/ensemble/_gradient_boosting.pyx:
```diff
@@ -112,25 +193,34 @@ def predict_stages(np.ndarray[object, ndim=2] estimators,
-                <float64 *> (<np.ndarray> out).data)
-            ## out += scale * tree.predict(X).reshape((X.shape[0], 1))
+    if issparse(X):
+        if not isinstance(X, csr_matrix):
```
I think this validation is unnecessary, seeing as all public entrypoints should already include a call to `check_array`. I'd rather not see ad-hoc error messages when `check_array` should make this consistent across the codebase.

jnothman commented:
Otherwise LGTM. Please rebase. Thanks for your attention to consistency, readability and efficiency.

olologin committed:
Refactoring of _gradient_boosting.pyx

ClimbsRocks commented:
@olologin great work here! any chance this has been finished? I'm currently running into issues with XGBoost and how they handle sparse matrices, so I was happy to switch over to sklearn's GradientBoostingRegressor, only to run into the issues you've outlined above. 

Getting this fixed would be awesome, and it seems you've already done most of the work  for it!

olologin commented:
@ClimbsRocks , It is finished. It just waits for review from scikit-learn team.

But you can merge this fix to your own branch and use it while it's not in master of upstream.

amueller commented:
This says +2. Who except @jnothman gave +1? @jaquesgrobler did you do a full review?

raghavrv commented on sklearn/ensemble/_gradient_boosting.pyx:
```diff
@@ -84,22 +88,96 @@ cdef void _predict_regression_tree_inplace_fast(DTYPE_t *X,
+
+    The function assumes that the ndarray that wraps ``X`` is csr_matrix.
+    """
+    cdef DTYPE_t* X_data = <DTYPE_t*>(<np.ndarray> X.data).data
```
Shouldn't it be `<np.ndarray[ndim=1, dtype=DTYPE_t]>`?

Could you refer [`_apply_sparse_csr`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/tree/_tree.pyx#L814) and refactor it the same way. I feel for an additional 3 lines, it will be clearer... :)

olologin replied:
Sorry, I added <np.ndarray[ndim=1, dtype=DTYPE_t]> to my code, and now it takes almost as much space as version from _apply_sparse_csr, so I don't see any sense in refactoring _apply_sparse_csr.

raghavrv commented:
+1 from me too. Thanks for the fix @olologin

ogrisel commented:
For some reason I cannot get my valgrind memcheck to work with python on my box. Instead I tried the following script to check for memory leaks:

```python
from sklearn.ensemble import GradientBoostingRegressor
import numpy as np
from scipy.sparse import rand
import sys
import psutil
from time import time

X = rand(100, 50, density=0.01, format='csr', dtype=np.float32,
         random_state=0)
y = X.toarray()[:, 0] + .01 * np.random.randn(100)

if len(sys.argv) > 1 and sys.argv[1] == 'dense':
    X = X.toarray()


gbr = GradientBoostingRegressor(n_estimators=100).fit(X, y)


duration = 0
p = psutil.Process()
for i in range(10000):
    if i % 1000 == 0:
        print(p.memory_info().rss / 1e6)
    t0 = time()    
    gbr.predict(X)
    duration += time() - t0
print("Total time: %0.3fs" % duration)
```

The memory usage is stable. However the prediction time on sparse data is 5 time slower than for the same data represented as a dense array. I wonder if this can be optimized.

ogrisel commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -1050,6 +1050,9 @@ def check_sparse_input(EstimatorClass, X, X_sparse, y):
                               auto.feature_importances_)
 
+    assert_array_equal(sparse.predict(X_sparse), dense.predict(X))
+    assert_array_equal(dense.predict(X_sparse), sparse.predict(X))
```
Please use `assert_array_almost_equal` to compare arrays of potentially floating point values (when the estimator is a regressor).

olologin replied:
I thought we have to demand for these arrays to be completely equal, not "almost". That's why I used it.

ogrisel replied:
For floating point values this can vary: for instance if there are upcasts from `float` to `double` you don't always get the exact same values depending on the compilers and architecture.

raghavrv commented on sklearn/ensemble/_gradient_boosting.pyx:
```diff
@@ -112,25 +190,31 @@ def predict_stages(np.ndarray[object, ndim=2] estimators,
+                ## out += scale * tree.predict(X).reshape((X.shape[0], 1))
 
 
-@cython.nonecheck(False)
```
Why is this removed?

olologin replied:
@raghavrv Look at glouppe's comment from Jun 21:
>By default, nonecheck=False. This is decorator can therefore be removed.

raghavrv commented on sklearn/ensemble/_gradient_boosting.pyx:
```diff
@@ -84,22 +88,96 @@ cdef void _predict_regression_tree_inplace_fast(DTYPE_t *X,
+    free(X_sample)
+    free(feature_to_sample)
+
+
```
I think you should try turning off the [cython checks](http://cython.readthedocs.io/en/latest/src/reference/compilation.html#compiler-directives)? It should speed it up a bit? Could you profile the speed with/without the checks and see if there is a speed improvement?

olologin replied:
@raghavrv, ok, I'll try to test it with turned off checks. But seems that most time-consuming checks are already disabled:

    # cython: cdivision=True
    # cython: boundscheck=False
    # cython: wraparound=False

It's actually interesting why sparse is 5 times slower, because I did my own benchmarks (Code is provided above), it was slower, but not even 2 times slower.

olologin committed:
Sparse prediction speed is faster now, also fixed safe_alloc usage

olologin committed:
assert_array_equal -> assert_array_almost_equal

olologin commented:
I fixed performance issue, now it works almost as fast as dense version in test provided by @ogrisel above.

Also I've found and fixed stupid mistake in safe_realloc usage from tree.pyx and in function for sparse prediction which I added here. It required more memory to allocate than user needs.

jnothman commented:
> This says +2. Who except @jnothman gave +1? @jaquesgrobler did you do a full review?

I can't see any cause for @olologin marking this as +1 before I incremented it. Changing back to +1.

jnothman commented on sklearn/ensemble/_gradient_boosting.pyx:
```diff
@@ -194,8 +209,8 @@ def predict_stages(np.ndarray[object, ndim=2] estimators,
-            raise ValueError("X should be in np.ndarray or csr_matrix format, got %s"
-                                     % type(X))
+            raise ValueError("X should be in np.ndarray or csr_matrix format,"
+              "got %s" % type(X))
```
indentation here

olologin commented:
I thought jaquesgrobler's
>Very useful fix this. I'm +1 for merging this once last points are addressed :+1:
Means +1, so I changed it to +1. Anyway, seems he didn't do full review.

raghavrv commented:
Thanks for the addressing... You also need to rebase to let the CIs run...

olologin commented:
@amueller, @ogrisel, Sorry, but could you ask someone for a review of this?

amueller commented:
@jmschrei maybe?

olologin committed:
indentation fix

jnothman commented:
This would indeed be nice to have in 0.18, but there's a good chance it's missed that boat.

jmschrei commented:
I'm travelling today. If you can wait until tomorrow I can get a review in.

amueller commented:
needs a rebase (and a review from @jmschrei ;)

jmschrei commented on sklearn/ensemble/_gradient_boosting.pyx:
```diff
@@ -84,22 +88,111 @@ cdef void _predict_regression_tree_inplace_fast(DTYPE_t *X,
+    cdef INT32_t* X_indices = <INT32_t*>(
+        <np.ndarray[ndim=1, dtype=INT32_t]> X.indices).data
+    cdef INT32_t* X_indptr = <INT32_t*>(
+        <np.ndarray[ndim=1, dtype=INT32_t]> X.indptr).data
```
I'm not against defining the number of dimensions and dtype, but does it actually do anything? I've never used it personally.

olologin replied:
Actually I don't know :), I saw this some Cython/Numpy tutorials, and I think it's always better to specify everything explicitly, when we want maximum performance.

jmschrei replied:
You're getting a pointer to the underlying pointer array anyway, why does it matter how you specify the numpy array? You're not using it at all. I know this is used in the sklearn codebase but I'm pretty sure from my own tests that it doesn't do anything.

olologin replied:
Fixed.

jmschrei commented on sklearn/ensemble/_gradient_boosting.pyx:
```diff
@@ -112,25 +205,31 @@ def predict_stages(np.ndarray[object, ndim=2] estimators,
+        _predict_regression_tree_stages_sparse(estimators, X, scale, out)
+    else:
+        if not isinstance(X, np.ndarray):
+            raise ValueError("X should be in np.ndarray or csr_matrix format,"
```
Good

jmschrei commented on sklearn/ensemble/_gradient_boosting.pyx:
```diff
@@ -84,22 +88,111 @@ cdef void _predict_regression_tree_inplace_fast(DTYPE_t *X,
+    cdef SIZE_t sample_i = 0
+    cdef SIZE_t feature_i = 0
+    cdef SIZE_t stage_i = 0
+    cdef SIZE_t output_i = 0
```
I don't think we set values for variables used as iterators in for loops in the remainder of the code.

olologin replied:
Again, I didn't check performance gain, but here:
https://cython.readthedocs.io/en/latest/src/quickstart/cythonize.html#typing-variables
It's said:
>Since the iterator variable i is typed with C semantics, the for-loop will be compiled to pure C code.

This is why I predefined them.

And additionaly, we have same variables in _tree.pyx.

jmschrei replied:
I'm saying that you don't need to set them to equal 0. This is something we removed from the tree code a few months ago, as it was unnecessary and implied that the initial value of 0 was meaningful.

olologin replied:
Ah, I understood what you mean. Done.

jmschrei commented on sklearn/ensemble/_gradient_boosting.pyx:
```diff
@@ -84,22 +88,111 @@ cdef void _predict_regression_tree_inplace_fast(DTYPE_t *X,
+    # which features are nonzero in the present sample.
+    cdef SIZE_t* feature_to_sample = NULL
+
+    safe_realloc(&X_sample, n_features)
```
So you're taking a sparse row and converting it to a dense row and then making a prediction on that? Do you think it might be a good idea to keep it in the sparse format and iterate down the tree that way? I guess it would cause there to be a linear time (with NNZ elements) scan for each node which would be extremely detrimental. I'm wondering about the extreme case where a single sample can't fit in memory, but is immensely sparse. I suppose this isn't a major issue.

olologin replied:
Yes, it's from _apply_sparse_csr.
@jmschrei, you are right :), but how would you check for existence of some particular features in particular row? For example if you use bin-search over `X_indices` it will take O(log(N)), where N is number of initialised features in current row. It takes way more time than O(1) access to value in array.

+ I don't think It's common to have such big feature space, even in sparse matrix. Implemented solution will work faster for 99% use cases.
+ Again, I tried to keep it consistent with _tree.pyx

jmschrei replied:
Okay this doesn't seem like an issue then.

jmschrei reviewed (approved):
Other than the few minor comments, this looks good to me. We probably need to go through all the safe_realloc's in the tree code and make sure we're not misallocating memory. I remember being confused if we needed to specify n or n*sizeof(dtype). That is probably another PR though.

jmschrei commented:
Now that I think about it actually, maybe it would be worth adding a test ensuring that both the dense and sparse versions run within a factor of 2 of each other? @amueller what is your position on timings in unit tests? I don't want this to deprecate in the future.

jnothman commented:
Timings in unit tests are problematic. Relative timings are going to be dependent on nonzero density in X, apart from architecture issues. I'm -1 for such tests, though it is worth benchmarking (on one architecture) at PR time to see that nothing crazy is happening.

olologin committed:
Minor changes

jmschrei commented:
LGTM. This has my +1.

jnothman commented on doc/whats_new.rst:
```diff
@@ -56,6 +56,11 @@ Bug fixes
      <https://github.com/scikit-learn/scikit-learn/pull/6178>`_) by `Bertrand
      Thirion`_
 
+   - :class:`ensemble.GradientBoostingClassifier` and :class:`ensemble.GradientBoostingRegressor`
```
I think this belongs under enhancements

jnothman commented:
Move the what's new and we'll merge. Thanks!

olologin committed:
what's new entry moved to enhancement section

olologin commented:
Thanks for review :+1:

Sandy4321 commented:
as was written
olologin commented on Aug 27, 2016 
I fixed performance issue, now it works almost as fast as dense version in test provided by @ogrisel above. 2.773s for dense and 3.104s for sparse.
Also I've found and fixed stupid mistake in safe_realloc usage from tree.pyx and in function for sparse prediction which I added here. It required more memory to allocate than user needs
if somebody may share a test case code

rth commented:
@Sandy4321 see https://github.com/scikit-learn/scikit-learn/issues/6101#issuecomment-390279695

Sandy4321 commented:
I see so your code looks like this
'''
from __future__ import print_function, division
import numpy as np

from sklearn.datasets import fetch_20newsgroups_vectorized
from sklearn.utils.validation import check_array

from sklearn.ensemble import GradientBoostingClassifier

data_train = fetch_20newsgroups_vectorized(subset="train")
data_test = fetch_20newsgroups_vectorized(subset="test")
X_train_sp = check_array(data_train.data, dtype=np.float32,
                      accept_sparse="csc")
checked_test = check_array(data_test.data, dtype=np.float32, accept_sparse="csr")
X_test_sp = checked_test[:3500, :]
y_train_sp = data_train.target
y_test_sp = data_test.target[:3500]

X_test_dense = X_test_sp.todense()

print("20 newsgroups")
print("=============")
print("X_train.shape = {0}".format(X_train_sp.shape))
print("X_train density = {0}"
      "".format(X_train_sp.nnz / np.product(X_train_sp.shape)))
print("y_train {0}".format(y_train_sp.shape))
print("X_test {0}".format(X_test_sp.shape))
print("X_test.format = {0}".format(X_test_sp.format))
print("X_test.dtype = {0}".format(X_test_sp.dtype))
print("y_test {0}".format(y_test_sp.shape))
print()

print("Classifier Training")
print("===================")
accuracy, test_time = {}, {}

name = "GradientBoostingClassifier_100_trees"
clf = GradientBoostingClassifier(n_estimators=100)
try:
    clf.set_params(random_state=0)
except (TypeError, ValueError):
    pass

print("Training %s ... " % name, end="")
clf.fit(X_train_sp, y_train_sp)

%timeit clf.predict(X_test_dense)
%timeit clf.predict(X_test_sp)
'''

Sandy4321 commented:
???
I put code between ''' and '''

Sandy4321 commented:
```

my code
```
like this?

Sandy4321 commented:
great it works!!!!!
thanks
